### PR TITLE
Export the directory

### PR DESCRIPTION
**Pull request by JoneLau**

### CODE REVIEW

**Comment by JoneLau** [comment]
update cocos-creator/einge-3d#245

File: Assets/Editor/Window.Export.cs [hunk 1]
```diff
@@ -69,27 +72,67 @@ public class GLTFExportWindow : EditorWindow {
         }
       }
 
-      this.reorderableList = new ReorderableList(this.scenes, typeof(SceneAsset), true, true, true, true);
-      this.reorderableList.drawHeaderCallback = (Rect rect) => {
+      this.list = new ReorderableList(this.scenes, typeof(SceneAsset), true, true, true, true);
```
Comment (jwu) [inline Assets/Editor/Window.Export.cs]: this.list ????

File: Assets/Editor/Window.Export.cs [hunk 2]
```diff
@@ -69,27 +72,67 @@ public class GLTFExportWindow : EditorWindow {
...
 
         EditorGUI.BeginChangeCheck();
         asset = EditorGUI.ObjectField(rect, asset as Object, typeof(SceneAsset), false);
         if (EditorGUI.EndChangeCheck()) {
-          this.reorderableList.list[index] = asset;
+          this.list.list[index] = asset;
+          this.jsonDirty = true;
+        }
+      };
+
```
Comment (jwu) [inline Assets/Editor/Window.Export.cs]: remove unused line

File: Assets/Editor/Window.Export.cs [hunk 3]
```diff
@@ -11,17 +11,18 @@
...
   public class GLTFExportWindow : EditorWindow {
     public string outputPath = "";
     public string projectName = "";
     public FileMode mode = FileMode.Mixed;
     public List<SceneAsset> scenes;
 
-    ReorderableList reorderableList = null;
+    ReorderableList sceneReorderableList = null;
+
+    public List<Object> directories;
```
Comment (jwu) [inline Assets/Editor/Window.Export.cs]: public 的变量都放到一起

File: Assets/Editor/Window.Export.cs [hunk 4]
```diff
@@ -11,17 +11,18 @@
...
-
 namespace exsdk {
   public class GLTFExportWindow : EditorWindow {
     public string outputPath = "";
     public string projectName = "";
     public FileMode mode = FileMode.Mixed;
     public List<SceneAsset> scenes;
 
-    ReorderableList reorderableList = null;
+    ReorderableList sceneReorderableList = null;
```
Comment (jwu) [inline Assets/Editor/Window.Export.cs]: uiScenes

File: Assets/Editor/Window.Export.cs [hunk 5]
```diff
@@ -11,17 +11,18 @@
...
+
+    public List<Object> directories;
+
+    ReorderableList directoryReorderableList = null;
```
Comment (jwu) [inline Assets/Editor/Window.Export.cs]: uiDirs

File: Assets/Editor/Exporter.cs [hunk 6]
```diff
@@ -110,6 +110,25 @@ public partial class Exporter {
...
+      List<Object> prefabs,
+      List<Object> modelPrefabs,
+      List<Mesh> meshes,
+      List<Texture> textures,
+      List<Texture> spriteTextures,
+      List<Material> materials,
+      List<Font> fonts
+      ) {
+      Dictionary<string, JSON_Asset> assetsJson = new Dictionary<string, JSON_Asset>();
+      Debug.Log("dest===" + dest);
```
Comment (jwu) [inline Assets/Editor/Exporter.cs]: unused code

File: Assets/Editor/Export.Directory.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            }
+          }
+          // TODO:animationClip
+        }
+      }
+
+      saveAssets(dest, prefabs, modelPrefabs, meshes, textures, new List<Texture>(), materials, fonts);
+    }
+
+    void walkDirectory(string path, DirWalkCallback fn) {
```
Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: walkDirectory => WalkDir

File: Assets/Editor/Export.Directory.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      string[] dirs = Directory.GetDirectories(path);
+      for (int i = 0; i < dirs.Length; i++) {
+        RecurseDirectory(dirs[i], fn);
+      }
+    }
+
+    void RecurseDirectory(string path, DirWalkCallback fn) {
```
Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: RecurseDirectory => RecurseDir

File: Assets/Editor/Export.Directory.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+using UnityEngine;
+using System.IO;
+using UnityEditor;
+
+using Newtonsoft.Json;
+
+using UnityEngine.UI;
+
+namespace exsdk {
+  public delegate bool DirWalkCallback(string path);
```
Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: WalkDirCallback

File: Assets/Editor/Export.Directory.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+  public partial class Exporter {
+
+    public void DirExec(string outputPath) {
```
Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: SaveDirs

Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: outputPath => dest, 传递进来的直接是 outputPath/resources

File: Assets/Editor/json-defines.cs [hunk 11]
```diff
@@ -14,6 +14,8 @@ public class JSON_ExportSettings {
     public string projectName = "";
     public FileMode mode = FileMode.Mixed;
     public List<string> scenes = new List<string>();
+
```
Comment (jwu) [inline Assets/Editor/json-defines.cs]: unused line

File: Assets/Editor/Export.Directory.cs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+      // ========================================
+
+      if (Directory.Exists(dest)) {
+        DirectoryInfo di = new DirectoryInfo(dest);
+        di.Delete(true);
+      }
+
+      Directory.CreateDirectory(dest);
+
+      var allDirectories = new List<string>();
```
Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: 统一 dirs 吧

File: Assets/Editor/Export.Directory.cs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+      var prefabs = new List<Object>();
+      var modelPrefabs = new List<Object>();
+      var meshes = new List<Mesh>();
+      var textures = new List<Texture>();
+      var spriteTextures = new List<Texture>();
+      var materials = new List<Material>();
+      var fonts = new List<Font>();
+
+      // get all export directories
+      for (int i = 0; i < directories.Count; i++) {
```
Comment (jwu) [inline Assets/Editor/Export.Directory.cs]: dirs

File: Assets/Editor/json-defines.cs [hunk 14]
```diff
@@ -14,6 +14,8 @@ public class JSON_ExportSettings {
...
+    public List<string> directories = new List<string>();
```
Comment (jwu) [inline Assets/Editor/json-defines.cs]: dirs

**Pull request merged**